crates/router/src/core/payment_methods.rs

```diff
@@ -959,6 +986,172 @@ pub async fn create_payment_method(
     Ok(services::ApplicationResponse::Json(response))
 }
 
+#[cfg(all(feature = "v2", feature = "payment_methods_v2"))]
+#[derive(Clone, Debug)]
+pub struct NetworkTokenPaymentMethodDetails {
+    network_token_requestor_reference_id: String,
+    network_token_locker_id: String,
+    network_token_pmd: Encryptable<Secret<serde_json::Value>>,
+}
+
+#[cfg(all(feature = "v2", feature = "payment_methods_v2"))]
+pub async fn network_tokenize_and_vault_the_pmd(
+    state: &SessionState,
+    payment_method_data: &pm_types::PaymentMethodVaultingData,
+    merchant_account: &domain::MerchantAccount,
+    key_store: &domain::MerchantKeyStore,
+    network_tokenization: Option<payment_methods::NetworkTokenization>,
+    network_tokenization_enabled_for_profile: bool,
+    customer_id: &id_type::GlobalCustomerId,
+) -> RouterResult<NetworkTokenPaymentMethodDetails> {
+    let is_network_tokenization_enabled_for_pm = match network_tokenization {
+        Some(network_tokenization) => network_tokenization.enabled,
+        None => false,
+    };
+
+    match (
```

Comment:
Do get value from match and do use?, to reducing nesting.

Validate other branches like,
```
let value = match condition {
 valid_branch => Ok(value),
 _ => Err,
}?
```